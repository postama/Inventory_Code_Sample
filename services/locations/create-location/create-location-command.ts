import { literal, intersection, type, TypeOf } from 'io-ts';
import { BaseLocationCommand } from '../shared/base-location-command';
import { validateType } from '../../../infrastructure/validator/validator';
import uuid from 'uuid/v4';

export const CreateLocationCommandValidator = intersection([
    BaseLocationCommand,
    type({ type: literal('CreateLocation') })
]);

export type CreateLocationCommand = TypeOf<typeof CreateLocationCommandValidator>;

export function createCreateLocationCommand(request: any): CreateLocationCommand {
    return validateType({ ...request, isActive: true, type: 'CreateLocation', locationId: uuid() }, CreateLocationCommandValidator);
}