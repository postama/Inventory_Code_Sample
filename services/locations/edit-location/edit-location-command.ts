import { intersection, type, literal, TypeOf } from 'io-ts';
import { BaseLocationCommand } from '../shared/base-location-command';
import { validateType } from '../../../infrastructure/validator/validator';

export const EditLocationCommandValidator = intersection([
    BaseLocationCommand,
    type({ type: literal('EditLocation') })
]);

export type EditLocationCommand = TypeOf<typeof EditLocationCommandValidator>;

export function createEditLocationCommand(request: any, locationId: string): EditLocationCommand {
    return validateType({ ...request, type: 'EditLocation', locationId }, EditLocationCommandValidator);
}