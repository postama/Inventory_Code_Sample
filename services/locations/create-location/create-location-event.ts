import { intersection, type, literal, TypeOf } from 'io-ts';
import { createLocation, LocationValidator } from '../../../repo/locations/location-type';
import { CreateLocationCommand } from './create-location-command';
import { validateType } from '../../../infrastructure/validator/validator';
import { createBaseEvent, BaseEventValidator } from '../../../repo/events/event-type';
import uuid from 'uuid/v4';

export const CREATE_LOCATION = 'LocationCreated';

export const LocationCreatedValidatorV1 = intersection([
    BaseEventValidator,
    type({
        type: literal(CREATE_LOCATION),
        version: literal(1),
        data: LocationValidator
    })
]);

export type LocationCreatedV1 = TypeOf<typeof LocationCreatedValidatorV1>;

export function createCreateLocationEvent(command: CreateLocationCommand): LocationCreatedV1 {
    const event = {
        ...createBaseEvent(command, uuid(), CREATE_LOCATION, 1),
        data: createLocation(command)
    };

    return validateType(event, LocationCreatedValidatorV1);
}