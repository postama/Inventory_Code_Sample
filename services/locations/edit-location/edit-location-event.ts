import { intersection, type, literal, TypeOf } from 'io-ts';
import { createLocation, LocationValidator } from '../../../repo/locations/location-type';
import { EditLocationCommand } from './edit-location-command';
import { validateType } from '../../../infrastructure/validator/validator';
import { createBaseEvent, BaseEventValidator } from '../../../repo/events/event-type';

export const EDIT_LOCATION = 'LocationEdited';

export const LocationEditedValidatorV1 = intersection([
    BaseEventValidator,
    type({
        type: literal(EDIT_LOCATION),
        version: literal(1),
        data: LocationValidator
    })
]);

export type LocationEditedV1 = TypeOf<typeof LocationEditedValidatorV1>;

export function createEditLocationEvent(command: EditLocationCommand): LocationEditedV1 {
    const event = {
        ...createBaseEvent(command, command.locationId, EDIT_LOCATION, 1),
        data: createLocation(command)
    };

    return validateType(event, LocationEditedValidatorV1);
}