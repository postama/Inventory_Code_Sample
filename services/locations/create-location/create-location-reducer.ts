import { Location } from '../../../repo/locations/location-type';
import { LocationCreatedV1 } from './create-location-event';

// tslint:disable-next-line variable-name
export function createLocationReducer(_location: Location, event: LocationCreatedV1): Location {
    return {
        ...event.data,
        siteId: event.siteId,
        created: event.created
    };
}