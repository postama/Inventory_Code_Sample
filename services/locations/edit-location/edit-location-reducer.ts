import { Location } from '../../../repo/locations/location-type';
import { LocationEditedV1 } from './edit-location-event';

export function editLocationReducer(location: Location, event: LocationEditedV1): Location {
    return {
        ...event.data,
        created: location.created
    };
}