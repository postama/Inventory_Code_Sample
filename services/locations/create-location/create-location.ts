import { APIGatewayEvent } from 'aws-lambda';
import { requestHandler } from '../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper';
import { CreateLocationCommand, createCreateLocationCommand } from './create-location-command';
import { checkUniqueCommand } from '../../../repo/events/event';
import { LocationCreatedV1, createCreateLocationEvent } from './create-location-event';
import { createLocationReducer } from './create-location-reducer';
import { save } from '../../../repo/locations/location';

export function entry(request: APIGatewayEvent) {
    return requestHandler(createLocation, request);
}

export async function createLocation(request): Promise<void> {
    const command: CreateLocationCommand = createCreateLocationCommand(request);
    await checkUniqueCommand(command);
    const event: LocationCreatedV1 = createCreateLocationEvent(command);
    event.state = createLocationReducer({} as any, event);
    await save(event);
    return;
}