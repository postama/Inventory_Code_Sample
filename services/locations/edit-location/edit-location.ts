import { APIGatewayEvent } from 'aws-lambda';
import { requestHandler } from '../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper';
import { EditLocationCommand, createEditLocationCommand } from './edit-location-command';
import { checkUniqueCommand } from '../../../repo/events/event';
import { LocationEditedV1, createEditLocationEvent } from './edit-location-event';
import { editLocationReducer } from './edit-location-reducer';
import { getById, save } from '../../../repo/locations/location';

export async function entry(request: APIGatewayEvent) {
    return requestHandler(editLocation, request, request.pathParameters.locationId);
}

export async function editLocation(request: any, locationId: string): Promise<void> {
    const command: EditLocationCommand = createEditLocationCommand(request, locationId);
    await checkUniqueCommand(command);
    const event: LocationEditedV1 = createEditLocationEvent(command);
    const location = await getById(command.customerId, locationId);
    event.state = editLocationReducer(location, event);
    await save(event);
    return;
}