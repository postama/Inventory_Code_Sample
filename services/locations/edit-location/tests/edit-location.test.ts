import uuid from 'uuid/v4';
import { createTestEditLocationCommand, createTestEditLocationEvent } from './test.util';
import { createEditLocationCommand } from '../edit-location-command';
import { createEditLocationEvent } from '../edit-location-event';
import { createTestLocation } from '../../../../repo/locations/test/test-util';
import { editLocationReducer } from '../edit-location-reducer';
import { entry, editLocation } from '../edit-location';
import { createTestCreateLocationEvent } from '../../create-location/tests/test.util';
import { save, getById } from '../../../../repo/locations/location';
import { LocationCreatedV1 } from '../../create-location/create-location-event';

import * as entryLib from '../../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper';
jest.mock('../../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper');
const entryMock = entryLib as jest.Mocked<typeof entryLib>;

describe('edit location', () => {
    describe('command structure validator', () => {
        test('should fail without user with a human readable error message', () => {
            const command = createTestEditLocationCommand({ user: undefined });
            expect.assertions(2);
            try {
                createEditLocationCommand(command, uuid());
            } catch (e) {
                expect(e.name).toBe('ValidationError');
                expect(e.message).toBe('Expecting string at user but instead got: undefined.');
            }
        });

        test('should pass with full structure', () => {
            const command = createTestEditLocationCommand();
            const validCommand = createEditLocationCommand(command, uuid());
            expect(validCommand).toMatchSnapshot({
                locationId: expect.any(String),
                customerId: expect.any(String),
                commandId: expect.any(String),
                siteId: expect.any(String)
            });
        });
    });

    describe('event creator', () => {
        test('should throw if given a bad command', () => {
            expect.assertions(2);
            try {
                const editLocationCommand = createTestEditLocationCommand({ user: undefined });
                createEditLocationEvent(editLocationCommand);
            } catch (e) {
                expect(e.name).toBe('ValidationError');
                expect(e.message).toBe('Expecting string at user but instead got: undefined.');
            }
        });

        test('should create an event if given a good command', () => {
            const editLocationCommand = createTestEditLocationCommand();
            const event = createEditLocationEvent(editLocationCommand);
            expect(event.type).toBe('LocationEdited');
        });
    });

    describe('reducer', () => {
        test('should edit a location from an event', () => {
            const location = createTestLocation();
            const event = createTestEditLocationEvent({ name: 'xyz' });
            const editedLocation = editLocationReducer(location, event);
            expect(editedLocation.name).toBe('xyz');
        });

        test('should not modify the created date of the location', () => {
            const location = createTestLocation();
            const event = createTestEditLocationEvent({ name: 'xyz' });
            const editedLocation = editLocationReducer(location, event);
            expect(editedLocation.created).toBe(location.created);
        });
    });

    describe('entry', () => {
        test('should call the entry handler', async () => {
            const command = createTestEditLocationCommand();
            await entry({ body: JSON.stringify(command), pathParameters: { itemId: uuid() } } as any);
            expect(entryMock.requestHandler).toBeCalled();
        });
    });

    describe('edit location - integration', () => {
        test('should successfully edit the location', async () => {
            const event: LocationCreatedV1 = createTestCreateLocationEvent();
            await save(event);
            const editCommand = createTestEditLocationCommand({ customerId: event.customerId, name: 'xyz' });
            await editLocation(editCommand, event.entityId);
            const item = await getById(event.customerId, event.entityId);
            expect(item.name).toBe('xyz');
        });
    });
});