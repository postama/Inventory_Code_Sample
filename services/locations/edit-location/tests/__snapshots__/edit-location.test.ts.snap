// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit location command structure validator should pass with full structure 1`] = `
Object {
  "commandId": Any<String>,
  "customerId": Any<String>,
  "isActive": true,
  "locationId": Any<String>,
  "name": "abc",
  "siteId": Any<String>,
  "type": "EditLocation",
  "user": "abc",
}
`;
